packages/s3-access-metrics/index.js

```diff
@@ -0,0 +1,178 @@
+// Given a stack name and a list of GET.OBJECT log events, return a list of
+// metric data objects for failure requests
+const getFailureMetricDataObjects = (stack, logEvents) => {
+  const groupedFailureEvents = groupBy(
+    logEvents.filter(isNotSuccessLogEvent),
+    getEventTimestampAsString
+  );
+  return Object.values(groupedFailureEvents)
+    .map(buildFailureMetricData(stack));
+};
+
+// Convert a list of log lines to a list of log events
+const logLinesToLogEvents = map(logLineToLogEvent);
+
+// Given an access log, return a list of log events
+const logEventsFromAccessLog = compose([logLinesToLogEvents, splitLines, trim]);
+
+// Given a stack name and an access log, return a list of metric data objects
+const metricDataObjectsFromAccessLog = curry((stack, accessLog) => {
+  const logEvents = logEventsFromAccessLog(accessLog);
+  const getObjectLogEvents = selectGetObjectLogEvents(logEvents);
+  return flatten([
+    getSuccessMetricDataObjects(stack, getObjectLogEvents),
+    getFailureMetricDataObjects(stack, getObjectLogEvents)
+  ]);
+});
+
+// Given a stack name and a logLocation object with Bucket and Key properties,
+// fetch that S3 Server Access Log from S3 and return a list of metric data
+// objects to be uploaded to Cloudwatch Metrics
+const getMetricDataObjectsFromAccessLog = curry((stack, logLocation) =>
+  fetchObject(logLocation)
+    .then(metricDataObjectsFromAccessLog(stack)));
+
+// Given a list of Cloudwatch Metric Data, upload that data to Cloudwatc
+const putMetricDataObjects = (metricDataObjects) => {
+  const cloudwatch = new AWS.CloudWatch();
+
+  const performPut = (MetricData) =>
+    cloudwatch.putMetricData({
+      Namespace: 'CumulusDistribution',
+      MetricData
+    }).promise();
+
+  const promisedPuts = chunk(metricDataObjects, 20).map(performPut);
+
+  return Promise.all(promisedPuts);
+};
+
+// Given a stack name and a list of S3 Server Access log locations, fetch the
+// logs and build metric data objects from the logs.
+const getMetricDataFromAccessLogs = (stack, logLocations) =>
+  Promise.all(logLocations.map(getMetricDataObjectsFromAccessLog(stack)))
+    .then(flatten);
+
+// Given an event record, return the S3 Bucket and Key that the event record
+// is referencing.
+const s3ParamsFromRecord = (record) => ({
+  Bucket: record.s3.bucket.name,
+  Key: record.s3.object.key
+});
+
+// Handle an S3 object upload event
+const handleEvent = (event, stack) => {
+  const logLocations = event.Records.map(s3ParamsFromRecord);
+
+  return getMetricDataFromAccessLogs(stack, logLocations)
+    .then(putMetricDataObjects);
+};
+
+module.exports = {
+  handler: async (event) => handleEvent(event, process.env.stack),
+
+  // Exported to support testing
+  buildMetricDataObject,
+  getHttpStatusFromLogLine,
+  getSuccessMetricDataObjects,
+  getOperationFromLogLine,
+  getTimestampFromLogLine,
+  getFailureMetricDataObjects,
+  logLineToLogEvent,
+  metricDataObjectsFromAccessLog,
+  s3ParamsFromRecord
```

Comment:
Would you rather use `rewire`?  Let's you reach into this file for testing without exporting.